runners/spark/src/main/java/org/apache/beam/runners/spark/util/CachedSideInputReader.java

```diff
@@ -86,9 +55,27 @@ private CachedSideInputReader(SideInputReader delegate) {
   @Override
   public <T> T get(PCollectionView<T> view, BoundedWindow window) {
     @SuppressWarnings("unchecked")
-    final Map<Key<T>, T> materializedCasted = (Map) materialized;
-    return materializedCasted.computeIfAbsent(
-        new Key<>(view, window), key -> delegate.get(view, window));
+    final Cache<Key<T>, Optional<T>> materializedCasted =
+        (Cache) SideInputStorage.getMaterializedSideInputs();
+
+    Key<T> sideInputKey = new Key<>(view, window);
+
+    try {
+      Optional<T> optionalResult =
+          materializedCasted.get(
+              sideInputKey,
+              () -> {
+                final T result = delegate.get(view, window);
+                LOG.info(
+                    "Caching de-serialized side input for {} of size [{}B] in memory.",
+                    sideInputKey,
+                    SizeEstimator.estimate(result));
+                return Optional.ofNullable(result);
+              });
+      return optionalResult.orElse(null);
```

Comment:
Ismaël is right. The `delegate.get` is not `@Nullable` at this place in the abstraction. You don't need to check for it (unless there's some other bug somewhere) and you shouldn't convert `Optional.absent()` to `null`.

Reply:
I chose this solution because guava `Cache` doesn't allow null values and I didn't realize I will break semantic meaning. I will try to find out different solution.

Reply:
Ah, thank you for clarifying. That is a good attempt. The problem is that these will incorrectly be turned into the same thing:

`Optional.ofNullable(null).orElse(null) == null`

`Optional.ofNullable(Optional.absent()).orElse(null) == null`

The fact that `Optional.of(null)` throws NPE is a mistake in the design (both Java and Guava). Maybe the point of the design is to convince people to not use `null`, which [is a billion dollar good idea](https://en.wikipedia.org/wiki/Null_pointer#History). But it makes `Optional<T>` not correctly parametric in `T`.

I think that if you actually convert `null` into `Optional.of(Optional.absent())` and other values `v` into `Optional.of(Optional.of(v))` you can simulate the behavior it should have had in the first place. Or you could make your own little replacement of `Optional`.

Reply:
Thanks for suggestions, Optional combo would be not very readable.
I made my own wrapper where I simply wrap the value so I can put `null` into the cache.
https://github.com/apache/beam/pull/7091/files#diff-b123f0f1ca9646966a641a458b74cfbcR92